app/Http/Controllers/DemandController.php

```diff
@@ -0,0 +1,334 @@
+
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\StreamedResponse;
+
+/**
+ * API commands used to serve banners and log relevant events
+ */
+class DemandController extends Controller
+{
+    public function serve(Request $request, $id)
+    {
+        $banner = Banner::find($id);
+        if (empty($banner)) {
+            abort(404);
+        }
+
+        // TODO:  ID / UUID here =>
+        // TODO: no need for obfuscation
+        // TODO: Yoda smell stuff here // this should be cleaned up
+
+        if ($request->getRealMethod() == 'OPTIONS') {
+            $response = new Response('', 204);
+        } else {
+            $response = new GzippedStreamedResponse();
+        }
+
+        if ($request->headers->has("Origin")) {
+            $response->headers->set("Access-Control-Allow-Origin", $request->headers->get("Origin"));
+            $response->headers->set("Access-Control-Allow-Credentials", "true");
+            $response->headers->set("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
+            $response->headers->set("Access-Control-Expose-Headers", "X-Adshares-Cid, X-Adshares-Lid");
+        }
+
+        if ($request->getRealMethod() == 'OPTIONS') {
+            $response->headers->set('Access-Control-Max-Age', 1728000);
+            return $response;
+        }
+
+        $isIECompat = $request->query->has("xdr");
+
+        if ($banner->creative_type == 'html') {
+            $mime = "text/html";
+        } else {
+            $mime = "image/png";
+        }
+
+        $tid = Utils::attachTrackingCookie(config('app.adserver_secret'), $request, $response, $banner->creative_sha1, $banner->updated_at);
+
+        $response->setCallback(function () use ($response, $banner, $isIECompat) {
+            if (!$isIECompat) {
+                echo $banner->creative_contents;
+                return;
+            }
+
+            $headers = [];
+            foreach ($response->headers->allPreserveCase() as $name => $value) {
+                if (strpos($name, "X-") === 0) {
+                    $headers[] = "$name:" . implode(",", $value);
+                }
+            }
+            echo implode("\n", $headers) . "\n\n";
+            echo base64_encode($banner->creative_contents);
+        });
+
+        $logIp = bin2hex(inet_pton($request->getClientIp()));
+        $cid = Utils::createTrackingId(config('app.adserver_secret'));
+
+        $log = new EventLog();
+        $log->banner_id =$banner->id;
+        $log->cid = Utils::getRawTrackingId($cid);
+        $log->tid = Utils::getRawTrackingId($tid);
+        $log->ip = $logIp;
+        $log->event_type = "request";
+        $log->save();
+
+        $response->headers->set('X-Adshares-Cid', $cid);
+        $response->headers->set('X-Adshares-Lid', $log->id);
+
+        if (! $response->isNotModified($request)) {
+            $response->headers->set('Content-Type', ($isIECompat ? 'text/base64,' : '') . $mime);
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "Content-Type" 4 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1192)